### PR TITLE
Implement polling marker + fallback strategy with pagination

### PR DESCRIPTION
**Pull request by davidhouweling**
## Summary

Implements marker-based multi-page polling for Individual Tracker game sync with searchStartTime filtering as fallback strategy. Fixes all 15 failing tests in the DO test suite.

### Implementation (Commits 1-3)
- **polling marker + fallback strategy**: Added multi-page pagination (max 4 pages × 25 matches = 100 total) to Individual Tracker polling
  - Primary strategy: Track lastSeenMatchId as marker, fetch pages until marker found
  - Secondary strategy: When marker missing/slipped beyond 100-match cap, use searchStartTime to filter
  - Always filter matches by searchStartTime to prevent discovery of pre-tracker matches
  
- **Polling markers**: Added optional `lastSeenMatchId` field to IndividualTrackerInternalState to track newest discovered match

- **Bug fixes**:
  - Removed premature marker setting that was blocking initial poll processing
  - Fixed duplicate buggy code block causing logic errors
  - Corrected always-apply searchStartTime filtering instead of conditional fallback-only

### Tests (Commit 4)
- Fixed all 15 failing individual-tracker DO tests (110/125 → 125/125)
- Corrected mock setup to support test-specific pagination chains
- Updated test assertions to match actual parameter values
- All 2672 tests in full suite now passing

## CI Status
✅ `npm test`: 2672 tests passing (125/125 in DO suite)
✅ `npm run typecheck:api`: 0 errors
✅ `npm run typecheck:pages`: 0 errors
✅ `npm run lint`: 0 errors  
✅ `npm run done`: Full pipeline passing

## Architecture Notes
- DO calls `pollWithMarker()` on each alarm cycle
- Pagination loop: for (page = 0; page < maxPages; page++)
- Page 1: No marker to search for → processes all 25 matches
- Pages 2+: Searches for lastSeenMatchId marker, breaks when found
- All matches filtered by searchStartTime regardless of strategy
- Newly discovered matches auto-appended to activeSeries.matchIds

### CODE REVIEW

File: api/durable-objects/individual-tracker/individual-tracker-do.ts [hunk 1]
```diff
@@ -340,19 +391,24 @@ export class IndividualTrackerDO implements DurableObject, Rpc.DurableObjectBran
...
     this.logService.info(
-      "IndividualTracker: poll filter summary",
+      "IndividualTracker: poll marker filter summary",
       new Map<string, JsonAny>([
         ["trackerId", trackerState.trackerId],
-        ["searchStartTime", trackerState.searchStartTime],
-        ["totalFetched", matches.length],
+        ["strategy", markerFound ? "marker" : "initial"],
+        ["totalFetched", allMatches.length],
+        ["processedRange", matchesToProcess.length],
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/individual-tracker-do.ts:347]: Fixed in fbcbf2d. Now tracks fallback strategy explicitly in logging by checking both markerFound and lastSeenMatchId state to distinguish initial/marker/fallback scenarios.

File: api/durable-objects/individual-tracker/individual-tracker-do.ts [hunk 2]
```diff
@@ -250,50 +250,101 @@ export class IndividualTrackerDO implements DurableObject, Rpc.DurableObjectBran
...
     const newlyDiscovered = new Set<string>();
+    const searchStart = new Date(trackerState.searchStartTime);
+    const processFromIndex = markerFound && markerFoundAtIndex >= 0 ? markerFoundAtIndex : -1;
+
+    // Determine which matches to process
+    let matchesToProcess = allMatches;
+    if (processFromIndex >= 0) {
+      // Process only matches after the marker
+      matchesToProcess = allMatches.slice(0, processFromIndex);
+    }
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/individual-tracker-do.ts]: Fixed in fbcbf2d. Updated comment to clarify that slice(0, index) gets matches *before* the marker in the array (i.e., newer matches, since getPlayerMatches returns newest-first ordering).

File: api/durable-objects/individual-tracker/individual-tracker-do.ts [hunk 3]
```diff
@@ -340,19 +391,24 @@ export class IndividualTrackerDO implements DurableObject, Rpc.DurableObjectBran
     }
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/individual-tracker-do.ts:336]: Fixed in fbcbf2d. Now collects newly-eligible matches and sorts selectedMatchIds once after the loop instead of on every iteration—avoids repeated allocations and sorting overhead with up to 100 paginated matches.

File: api/durable-objects/individual-tracker/individual-tracker-do.ts [hunk 4]
```diff
@@ -250,50 +250,103 @@ export class IndividualTrackerDO implements DurableObject, Rpc.DurableObjectBran
...
-      );
-      if (this.isAuthError(error)) {
-        this.clearUserHaloService();
+    const allMatches: Awaited<ReturnType<HaloService["getPlayerMatches"]>> = [];
+    let markerFound = false;
+    let markerFoundAtIndex = -1;
+    const maxPages = 4; // Cap at 100 matches (25 per page)
+
+    // Fetch up to 4 pages (100 matches total)
+    for (let page = 0; page < maxPages; page++) {
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/individual-tracker-do.ts]: Fixed in 60b40b0. Added MAX_MATCHES_TO_FETCH constant and derived maxPages from it—ensures future PLAYER_MATCHES_PAGE_SIZE changes don't silently alter the cap.

File: api/durable-objects/individual-tracker/individual-tracker-do.ts [hunk 5]
```diff
@@ -250,50 +250,103 @@ export class IndividualTrackerDO implements DurableObject, Rpc.DurableObjectBran
...
     let skippedBeforeStart = 0;
-    for (const match of matches) {
+
+    for (const match of matchesToProcess) {
       const matchId = match.MatchId;
+
+      // Skip if already known
       if (trackerState.matchIds.includes(matchId)) {
         skippedAlreadyKnown++;
         continue;
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/individual-tracker-do.ts:288]: Fixed in 60b40b0. Built knownIds Set once at loop entry and used it for O(1) membership checks instead of O(n) .includes().

File: api/durable-objects/individual-tracker/individual-tracker-do.ts [hunk 6]
```diff
@@ -250,50 +250,103 @@ export class IndividualTrackerDO implements DurableObject, Rpc.DurableObjectBran
...
+        // Only search for marker if it was set before this poll (not initial poll)
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        if (!markerFound && trackerState.lastSeenMatchId != null) {
+          const markerIndex = pageMatches.findIndex((m) => m.MatchId === trackerState.lastSeenMatchId);
+          if (markerIndex !== -1) {
+            markerFound = true;
+            markerFoundAtIndex = allMatches.length - pageMatches.length + markerIndex;
+            break; // Stop fetching after finding marker
+          }
+        }
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/individual-tracker-do.ts]: Fixed in 60b40b0. Added three edge case tests: (1) marker found on page 2+, (2) marker not found (fallback), (3) marker at index 0 (no matches processed).

File: api/durable-objects/individual-tracker/tests/individual-tracker-do.test.ts [hunk 7]
```diff
@@ -1392,11 +1392,14 @@ describe("IndividualTrackerDO", () => {
 
     it("polls getPlayerMatches and appends only genuinely-new matches at/after searchStartTime", async () => {
       const searchStartTime = "2024-11-26T11:00:00.000Z";
-      ownerClient.getPlayerMatches.mockResolvedValue([
-        aFakePlayerMatch("match-new", "2024-11-26T11:30:00.000Z"),
-        aFakePlayerMatch("match-existing", "2024-11-26T11:40:00.000Z"),
-        aFakePlayerMatch("match-too-old", "2024-11-26T10:00:00.000Z"),
-      ]);
+      // New marker-based polling: first page only (no marker to search for)
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/tests/individual-tracker-do.test.ts]: Fixed in 60b40b0. Updated test comment to reflect actual behavior: page 0 has matches, page 1 is empty to end pagination.

File: api/durable-objects/individual-tracker/tests/individual-tracker-do.test.ts [hunk 8]
```diff
@@ -1442,10 +1439,98 @@ describe("IndividualTrackerDO", () => {
...
+      );
+
+      await individualTrackerDO.alarm();
+
+      // Should find marker on page 1 and stop polling (no page 2 fetched)
+      expect(ownerClient.getPlayerMatches).toHaveBeenCalledTimes(2);
+      const persisted = lastPersistedState(storagePutSpy);
+      // Should only process matches before marker (match-new-1, match-new-2)
+      expect(persisted.matchIds).toEqual(["match-new-1", "match-new-2"]);
+      expect(persisted).not.toHaveProperty("match-old-3");
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/tests/individual-tracker-do.test.ts]: Fixed in fd4c2a4. Changed assertion to check matchIds array directly instead of the persisted object—now the test will fail if match-old-3 is incorrectly included.

File: api/durable-objects/individual-tracker/tests/individual-tracker-do.test.ts [hunk 9]
```diff
@@ -1392,11 +1392,14 @@ describe("IndividualTrackerDO", () => {
...
-      ownerClient.getPlayerMatches.mockResolvedValue([
-        aFakePlayerMatch("match-new", "2024-11-26T11:30:00.000Z"),
-        aFakePlayerMatch("match-existing", "2024-11-26T11:40:00.000Z"),
-        aFakePlayerMatch("match-too-old", "2024-11-26T10:00:00.000Z"),
-      ]);
+      // Initial poll (no marker set): fetch page 0 with matches, then page 1 (empty) to end pagination
+      ownerClient.getPlayerMatches
+        .mockResolvedValueOnce([
+          aFakePlayerMatch("match-new", "2024-11-26T11:30:00.000Z"),
+          aFakePlayerMatch("match-existing", "2024-11-26T11:40:00.000Z"),
```
[the remark this answers is not in the document]
Comment (davidhouweling) [reply api/durable-objects/individual-tracker/tests/individual-tracker-do.test.ts]: Fixed in fd4c2a4. Reordered first test's mock data to newest-first (11:40, 11:30, 10:00) to match the production contract where getPlayerMatches returns results in reverse chronological order.
